rust-symcrypt/src/hash.rs

```diff
@@ -0,0 +1,186 @@
+use core::ffi::c_void;
+use std::mem;
+use std::ptr;
+use symcrypt_sys;
+
+pub const SHA256_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
+pub const SHA384_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
+
+pub struct Sha256State {
+    state: symcrypt_sys::SYMCRYPT_SHA256_STATE,
+}
+
+impl Sha256State {
+    pub fn new() -> Self {
+        let mut instance = Sha256State {
+            state: symcrypt_sys::SYMCRYPT_SHA256_STATE::default(),
+        };
+        unsafe {
+            symcrypt_sys::SymCryptSha256Init(&mut instance.state);
+        }
+        instance
+    }
+
+    pub fn append(&mut self, data: &[u8]) {
+        unsafe {
+            symcrypt_sys::SymCryptSha256Append(
+                &mut self.state,                    // pState
+                data.as_ptr(),                      // pbData
+                data.len() as symcrypt_sys::SIZE_T, // cbData
+            );
+        }
+    }
+
+    pub fn result(&mut self, result: &mut [u8; SHA256_RESULT_SIZE]) {
+        unsafe {
+            symcrypt_sys::SymCryptSha256Result(&mut self.state, result.as_mut_ptr());
+        }
+    }
+}
+
+impl Drop for Sha256State {
+    fn drop(&mut self) {
+        unsafe {
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.state) as *mut c_void,
+                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
+            )
+        }
+    }
+}
+
+pub fn sha256(data: &[u8], result: &mut [u8; SHA256_RESULT_SIZE]) {
+    unsafe {
+        symcrypt_sys::SymCryptSha256(
+            data.as_ptr(),                      // pbData
+            data.len() as symcrypt_sys::SIZE_T, // cbData
+            result.as_mut_ptr(),                // pbResult
+        );
+    }
+}
+
+pub struct Sha384State {
+    state: symcrypt_sys::SYMCRYPT_SHA384_STATE,
+}
+
+impl Sha384State {
+    pub fn new() -> Self {
+        let mut instance = Sha384State {
+            state: symcrypt_sys::SYMCRYPT_SHA384_STATE::default(),
+        };
+        unsafe {
+            symcrypt_sys::SymCryptSha384Init(&mut instance.state);
+        }
+        instance
+    }
+
+    pub fn append(&mut self, data: &[u8]) {
```

Comment:
Since this has basically the same methods as `Sha256State`, it might be worth pulling them into a trait like `HashState`. You could use associated types on that to account for the different result types.

Reply:
Fixed via @eholk's commit